python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py

```diff
@@ -74,6 +77,19 @@ def resolve_dependencies(self, _graphene_info):
             for dep in self._external_asset_node.dependencies
         ]
 
+    def resolve_inProgressRuns(self, _graphene_info):
+        runs = []
+        for job_name in self._external_asset_node.job_names:
+            runs.extend(
+                [
+                    GrapheneRun(run)
+                    for run in get_in_progress_runs_with_in_progress_step(
```

Comment:
Hmm, I think this is fine but I worry making `inProgressRuns` a leaf node resolver is causing us to do more work than necessary? I think we'll be calling `get_in_progress_runs` for every asset node, and then `get_run_step_stats` for every run - in a 50 node graph with 3 active runs I think we'd be talking ~300 postgres queries in this code path?

The `get_in_progress_runs` query in particular seems like it would benefit from being on the outside of the loop, since the instance probably only has a handful of in progress jobs and this call might return the same small set of jobs for every node.

I wonder if it'd be faster to flip this over and load every in-progress job, then pass the array of asset node ops to `get_run_step_stats` once for each run, and then collate those into a result set by asset node? It'd mean a separate GraphQL query we'd have to merge client-side, but that'd be ok.

(I don't know that it's worth reworking what you've got, I've just been a bit frustrated with these fan out resolvers elsewhere so I'm looking for them now  👀  We could also try to resolve this more generally by letting gunicorn paralellize resolvers somehow and using a postgres connection pool maybe?)

Reply:
@bengotow Yes, this makes a lot of sense. I modified the query and added an `InProgressRunsByStep` field containing all the in-progress runs separated by asset node key to the `GrapheneRepository` object. Let me know if this is what you were looking for!